### PR TITLE
gradle: build cache tweaks

### PR DESCRIPTION
**Pull request by postamar**
The performance of `.github/workflows/gradle.yml` leaves much to be desired, so I looked into a recent scan: https://scans.gradle.com/s/g7y5vlevtbui6/performance/build-cache

## cache performance

The build cache hit rate is pitiful at ~3% and I noticed that pushing to the S3 cache was disabled in this workflow. It was enabled only when in airbyte-ci, i.e. connector builds and tests, but those build cache contents fall under a different prefix anyway, so I'm enabling it in all CI runs so that the gradle.yml workflow can use it properly.

This PR enables pushing to the local build cache in CI as well because it may actually be indirectly persisted via the burrunan/gradle-cache-action@v1 action. In any case it won't hurt to do this in CI.

I ran a test with this PR:
1. I had the workflow run once with these changes to actually warm the cache (by pushing). 
2. I ran it again but adding some noice in a CDK package to trigger the CDK tests to simulate a real PR run where the cache is invalidated to a reasonable degree. The cache hit rate is now 88%. See scan: https://scans.gradle.com/s/sin77pqyg2dfe/performance/build-cache
3. I ran it yet again without invalidating the cache (simply by clicking the "re-run all jobs" button in github actions). The cache hit rate is now 96%. See scan: https://scans.gradle.com/s/ytwacksbqp3fm/performance/build-cache

Wall-clock workflow latency used to be in the 5-minute range prior to this PR. By fixing the cache it went down to 3-4 minutes. 

## parallelization and runner size

Gradle scans also display a Gantt chart timeline and report a "serial execution factor". It went from ~32x originally down to ~12x in runs (2) and (3). This should mean that we don't need to run this on a 64-core machine and a 16-core machine will do. Unfortunately the S3 build cache sometimes randomly stops working so we need to actually reduce the workload in case the cache can't be used otherwise the workflow latency goes up to 12 minutes in those cases (I tested).

A big part of the workload is running static analysis with Spotbugs. A lot of the code covered is testing code. This seems wasteful. I tuned it to skip test files in this workflow.

### CODE REVIEW

File: .github/workflows/gradle.yml [hunk 1]
```diff
@@ -37,17 +37,18 @@ jobs:
             java:
               - '**/*.java'
               - '**/*.gradle'
+              - '**/*.kt'
```
Comment (postamar) [inline .github/workflows/gradle.yml:40]: we forgot about kotlin!

File: .github/workflows/gradle.yml [hunk 2]
```diff
@@ -37,17 +37,18 @@ jobs:
...
-    # The gradle check task which we will run is embarrassingly parallelizable.
-    # We therefore run this on a machine with a maximum number of cores.
-    # We pay per time and per core, so there should be little difference in total cost.
-    # The latency overhead of setting up gradle prior to running the actual task adds up to about a minute.
-    runs-on: connector-test-xxlarge
+    # As of now, 16 cores seems to be enough.
+    # Any revision upwards should be based on a performance analysis of gradle scans.
+    # See https://github.com/airbytehq/airbyte/pull/36055 for an example of this,
+    # which explains why which we went down from 64 cores to 16.
+    runs-on: connector-test-large
```
Comment (postamar) [inline .github/workflows/gradle.yml:51]: see PR description for why this is OK
tl;dr better caching, disable static analysis on test binaries

File: build.gradle [hunk 3]
```diff
@@ -98,7 +98,7 @@ allprojects {
 
     spotbugs {
         ignoreFailures = false
-        effort = Effort.valueOf(System.getProperty('skipSlowTests', 'false') == 'false' ? 'MAX' : 'MIN')
+        effort = Effort.valueOf('MAX')
```
Comment (postamar) [inline build.gradle:101]: this shouldn't cost too much with caching working properly

File: build.gradle [hunk 4]
```diff
@@ -209,12 +208,17 @@ allprojects {
...
         reports {
             xml.required = false
             html.required = true
         }
     }
 
-    javadoc.options.addStringOption('Xdoclint:none', '-quiet')
+    javadoc {
+        options.addStringOption('Xdoclint:none', '-quiet')
+    }
```
Comment (postamar) [inline build.gradle:223]: stylistic change only

File: build.gradle [hunk 5]
```diff
@@ -209,12 +208,17 @@ allprojects {
     }
 
     tasks.withType(SpotBugsTask).configureEach {
+        if (name != "spotbugsMain" && System.getProperty('skipSlowTests', 'false') != 'false') {
+            enabled = false
+        }
```
Comment (postamar) [inline build.gradle:213]: disables static analysis on test binaries

File: settings.gradle [hunk 6]
```diff
@@ -19,11 +19,10 @@ pluginManagement {
 // as much information as possible.
 plugins {
     id "com.gradle.enterprise" version "3.15.1"
-    id 'com.github.burrunan.s3-build-cache' version "1.5"
+    id 'com.github.burrunan.s3-build-cache' version "1.8.1"
```
Comment (postamar) [inline settings.gradle:22]: perhaps this helps?

File: settings.gradle [hunk 7]
```diff
@@ -115,27 +114,30 @@ gradleEnterprise {
...
 
-if (isCiServer || isAirbyteCI) {
+if (isCI) {
     buildCache {
         local {
             // Local build cache is dangerous as it might produce inconsistent results
             // in case developer modifies files while the build is running
-            enabled = isAirbyteCI
+            enabled = true
+            push = true
```
Comment (postamar) [inline settings.gradle:130]: this was missing, don't know if it's useful, but it doesn't hurt

**Review by stephane-airbyte — APPROVED** [review]
